src/constructs/iam/fastly-logs-iam.ts

```diff
@@ -0,0 +1,68 @@
+import { AccountPrincipal } from "@aws-cdk/aws-iam";
+import type { GuStack } from "../core";
+import { GuFastlyCustomerIdParameter } from "../core/parameters/fastly";
+import { GuPutS3ObjectsPolicy, GuRole } from "./";
+
+export interface GuFastlyLogsIamProps {
+  /**
+   * S3 bucket name that Fastly will ship logs
+   */
+  bucketName: string;
+  /**
+   * Path within the S3 bucket where Fastly will ship logs.
+   * @default - '*'
+   * If path is not specified, access will be granted to the entire S3 bucket.
+   */
+  path?: string;
+}
+
+// Fastly's AWS account ID is used as an external ID when creating the IAM role
+// See https://docs.fastly.com/en/guides/creating-an-aws-iam-role-for-fastly-logging
+const FASTLY_AWS_ACCOUNT_ID = "717331877981";
+
+/**
+ * Construct which creates the required IAM resources to support Fastly logging to an S3 bucket.
+ * Importantly, it does not create a permanent IAM user, which was once a requirement.
+ *
+ * Specifically, this construct creates a role which can only be assumed in Fastly's AWS account,
+ * with an attached policy allowing PUT access to an S3 bucket and path specified in the props.
+ *
+ * As mentioned, you should configure the S3 bucket used with `bucketName` and the path within
+ * that bucket with the optional `path`. `path` defaults to `"*"` i.e. access is given to the
+ * entire bucket.
+ *
+ * ```typescript
+ * new GuFastlyLogsIam(stack, "FastlyS3Logging", {
+ *   bucketName: "gu-mobile-logs"
+ *   path: "fastly/*"
+ * })
+ * ```
+ *
+ * You may instead wish to create separate roles, rather than a single one.
+ *
+ * ```typescript
+ * new GuFastlyLogsIam(stack, "FastlyS3LoggingMyApp", {
+ *   bucketName: "gu-mobile-logs"
+ *   path: "fastly/my-app/PROD/*"
+ * })
+ * ```
+ *
+ * See https://docs.fastly.com/en/guides/creating-an-aws-iam-role-for-fastly-logging
+ */
+export class GuFastlyLogsIam {
```

Comment:
I think this should be extend `GuRole` rather than being a plain class.

Reply:
I've made that change, although it doesn't make a whole lot of sense now because I have a policy being created inside a role constructor. I think semantically it was correct before - a construct formed of 2 resources (policy + role) which both used `Gu` constructs themselves (but the containing construct was its own entity).

Tempted to revert to previous implementation now!

Reply:
Extending `GuRole` follows the current convention of the repo, for example the [`GuInstanceRole`](https://github.com/guardian/cdk/blob/bf89943f4ea655994088870d9caa4e08bd5e2089/src/constructs/iam/roles/instance-role.ts#L36). This also makes me realise that this should probably be called `GuFastlyLogRole` too!

To nitpick a little:

> I think semantically it was correct before - a construct formed of 2 resources

This isn't correct as before we had a _vanilla class_ formed of two resources. In `CDK` there is a [`Construct` class](https://docs.aws.amazon.com/cdk/api/v1/docs/@aws-cdk_core.Construct.html) which almost everything extends from 🤯 .

I think generally we'd only create vanilla classes for patterns, rather than constructs - but even then, I think patterns could extend `Construct` (see [this example](https://github.com/aws-samples/aws-cdk-examples/blob/master/typescript/static-site/static-site.ts#L24) of a pattern from AWS).

One of the benefits from using the `Construct` type is you can hook into `validate` more easily. Additionally any exceptions that you might throw in the constructor of a `Construct` get pretty printed with a breadcrumb so it's a little easier to debug.